Nagios Services

adaniels21487 opened the pull request:
Hi Team,
This PR has some modifications to Nagios Services.

- Moved check-services.php to a poller module
- Added performance data collection to the poller
- Centralised DB functions to includes/services.inc.php
- Created add/edit/delete functions off the device, services page
- Removed legacy edit/delete interfaces.
- Moved and modified check.inc scripts
- Documentation Updates
- Modified services table

Some Notes...

1. Crontab entry no longer required.
As this is now a poller module, the cron entry for check-services.php is no longer required.
The documentation has been updated but what to do where this is already deployed in the field.

2. Add/Edit/Delete Pages
Service Add/Edit/Delete pages now exist on the device, services page.
But the device, services link doesn't exist until a service is created.
This means users must add the first service to a device using the narbar 'Add Service', or Device -> Edit and from then on services on this device can be managed within device -> services.
I'm not overly happy with this but I don't know how else to go. If anyone has a suggestion on how to handle this better, Please let me know.

3. Old check.inc scripts
These files have been moved up one level and named according to the check. eg. Old: includes/services/icmp/check.inc, New: includes/services/check_icmp.inc.php
These files used to only run the script, they now populate several variables that are used by the services system. See includes/services/check_icmp.inc.php for an example.
  - $check_cmd - builds the command to be run to execute the check
  - $check_ds - a replacement json array for the one that is discovered, this is used when you want to combine multiple DS's into a single graph. See 4.
  - $check_graph - the graph definitions when DS's are combined.


4. Performance data collection
The collection of performance data is now dynamic. When the service is polled DSs are extracted from the returned data. RRDs are created and DSs are saved in the database for graph creation.
There are some instances where the check script will return multiple DSs that could be combined into a single graph. For example ICMP, this check returns: rta, rtmin and rtmax. 
This can be done by creating a check script in includes/services/ for the type.

Please take a look and let me know what you think.

Aaron

adaniels21487 commented:
resolves #2193.

f0o commented on html/includes/forms/parse-service.inc.php:
```diff
@@ -0,0 +1,33 @@
+    die('ERROR: You need to be admin');
+}
+
+$service_id = $_POST['service_id'];
```
if possible use `$vars['service_id'];`

f0o commented on html/pages/addsrv.inc.php:
```diff
@@ -8,9 +8,7 @@
-            // FIXME should call add_service (needs more parameters)
-            $service_id = dbInsert(array('device_id' => $_POST['device'], 'service_ip' => $_POST['ip'], 'service_type' => $_POST['type'], 'service_desc' => $_POST['descr'], 'service_param' => $_POST['params'], 'service_ignore' => '0', 'service_status' => '0', 'service_checked' => '0', 'service_changed' => '0', 'service_message' => 'New check', 'service_disabled' => '0'), 'services');
-
+            $service_id = service_add($_POST['device'], $_POST['type'], $_POST['descr'], $_POST['ip'], $_POST['params'], 0);
```
also here please `$vars` if possible

f0o commented on html/pages/device/edit/services.inc.php:
```diff
@@ -3,30 +3,24 @@
-        echo 'yeah';
-        if ($_SESSION['userlevel'] >= '10') {
-            include 'includes/service-edit.inc.php';
+            $service_id = service_add($_POST['device'], $_POST['type'], $_POST['descr'], $_POST['ip'], $_POST['params'], 0);
```
`$vars` please

adaniels21487 commented:
Thanks @f0o, all fixed.

laf commented:
I don't use the services checks much but a concern would be around putting this into polling. I would have thought you would have a potential higher chance that these service checks could delay polling considerably. Using things like check_pop, check_smtp, check_http etc all have quite high timeout values. If things started failing these checks could take longer leading to polling not completing in time.

@librenms/reviewers thoughts?

Great work though, much needed love for this module :)

ciscoqid commented:
On 3/15/2016 4:35 PM, Neil Lathwood wrote:
>
> I don't use the services checks much but a concern would be around putting this into
> polling. I would have thought you would have a potential higher chance that these
> service checks could delay polling considerably. Using things like check_pop,
> check_smtp, check_http etc all have quite high timeout values. If things started failing
> these checks could take longer leading to polling not completing in time.
>
> @librenms/reviewers <https://github.com/orgs/librenms/teams/reviewers> thoughts?
>

Would it make sense to have a separate polling thread set for this so the SNMP threads are
protected? I can think of numerous ways to do that, but perhaps some sort of tagging ala
Ansible, so different scheduled pollers are restricted to those tags?  Tagging may be too
much -- could probably just be "time-sensitive" and "other" :).

Mark

-- 
Mark D. Nagel, CCIE #3177 Emeritus <mnagel@willingminds.com>
Principal Consultant, Willing Minds LLC (http://www.willingminds.com)
cell: 949-279-5817, desk: 714-495-4001, fax: 714-844-4698

** For faster support response time, please
** email support@willingminds.com or call 714-495-4000

laf commented:
I'm not sure it's of any benefit in tying this into the existing polling system Mark.

adaniels21487 commented:
Hi Team,
I have a bit on at work this week, I'll move this back to a standalone script over the weekend.

Aaron

paulgear commented:
I've merged this into my laptop branch for testing.    Here's a screenshot:
![screenshot from 2016-03-20 16-45-11](https://cloud.githubusercontent.com/assets/107222/13903168/7ceab8a8-eebb-11e5-805b-3396b1648fa5.png)

I think putting it back in a separate check script would still be a good idea, but this is a huge step forward.  The biggest thing I like about this is the flexibility we gain by graphing the performance data returned by Nagios.

I think the following improvements could be made (but they don't necessarily need to be in this PR):
- allow editing & deleting services from the master list of services
- when you click the name of the device in the service list, it should take you to the service screen for that device, not the overview screen
- there's a minor display bug in the graph legend: the line ends in \
- would be nice if the service list were sorted; I had a branch ready to go with this but decided to wait until this PR is merged before submitting
- in the longer term we need to find a good way to:
  1. add the same check to all/many devices at once
  1. select services for automatic enabling at discovery time (e.g. I want all my devices to be checked for an HTTP service, and if it's there, monitor it with this default alert)

From a coding perspective:
- The renames of discover_service() to service_discover() (and the other requivalents) are counter-intuitive; function names should be or begin with verbs.
- Eliminate the extra files in includes/services/check_*.inc.php and just use a config array (or better still, database entry).  I think we should make this as configurable as possible by the end user.  Ideally, when we add a service check, we should default to using the parameters which were used when the last check of the same type was added, so that people can just use any Nagios check they want without us having to explicitly support it.  (Maybe we could allow a custom path for the check rather than requiring them all to be in one place?)
- We should fix "TODO: Use proper Nagios service status. 0=Ok,1=Warning,2=Critical,Else=Unknown" before we release this (this was one of my pet peeves with the current services code).  The values in use by the state management stuff that @Rosiak has been working on are modelled directly on Nagios, and we should integrate with that.  Not many people are doing service polling now, and we should make the switch sooner rather than later, because this PR will cause a lot more people to be interested in service alerting.  A 1-2 week lead time should be enough for people to get their alert rules fixed.

adaniels21487 commented:
Hi @paulgear,
Thanks for the detailed review.

> I think the following improvements could be made (but they don't necessarily need to be in this PR):
> 
> - allow editing & deleting services from the master list of services

Good idea, it will be done.
> - when you click the name of the device in the service list, it should take you to the service screen for that device, not the overview screen

it will be done.
> - there's a minor display bug in the graph legend: the line ends in \

it will be fixed.
> - would be nice if the service list were sorted; I had a branch ready to go with this but decided to wait until this PR is merged before submitting

Which field would you like this sorted on?
> in the longer term we need to find a good way to:
> - add the same check to all/many devices at once
> - select services for automatic enabling at discovery time (e.g. I want all my devices to be checked for an HTTP service, and if it's there, monitor it with this default alert)

Tricky one, LibreNMS doesnt have the templating stuff other NMS's have as we have gone the autodiscovery route. Could the service discovery module assist here, I havent used it but it should add everything.
> From a coding perspective:
> 
> - The renames of discover_service() to service_discover() (and the other requivalents) are counter-intuitive; function names should be or begin with verbs.

You are correct, I dont know why I did this, it will be sorted, I also thought of wrapping this in a PHP class. Thoughts?
> - Eliminate the extra files in includes/services/check_*.inc.php and just use a config array (or better still, database entry). I think we should make this as configurable as possible by the end user. Ideally, when we add a service check, we should default to using the parameters which were used when the last check of the same type was added, so that people can just use any Nagios check they want without us having to explicitly support it. (Maybe we could allow a custom path for the check rather than requiring them all to be in one place?)

These can't move to the config array as-is as the $check_cmd is built from variables that dont exist until these services code is run.
I was looking to drop the 'IP Address' field and have all areguments in the parameters field. This would allow us to drop $check_cmd altogether, but I would need a way to run php on upgrade to modify all this.
I was then going to add a tooltip to each command to assist the user. I do like the idea of a separate path per check.
> - We should fix "TODO: Use proper Nagios service status. 0=Ok,1=Warning,2=Critical,Else=Unknown" before we release this (this was one of my pet peeves with the current services code). The values in use by the state management stuff that @Rosiak has been working on are modelled directly on Nagios, and we should integrate with that. Not many people are doing service polling now, and we should make the switch sooner rather than later, because this PR will cause a lot more people to be interested in service alerting. A 1-2 week lead time should be enough for people to get their alert rules fixed.

I was looking at bringing services into the component feature (which I am doing a similar state thing with) but to do this I need to run some php on upgrade to move all the data around. I can't find any doco, is there some reference code for the state system?

adaniels21487 commented:
> there's a minor display bug in the graph legend: the line ends in \

@paulgear, can you please send me the rrd command. I can see the slash in your screenshot but can find it.

laf commented:
@adaniels21487 We have no way of running php code on update at present, the only times we've needed to move data around it's been rrd files so that's just been done in the relevant poller file.

You could do it in includes/sql-schema/update.php after the schema run. That's probably the best place as it's either called manually, from daily or discovery (so you'd need to have a check if it needs to run as it would just continue to be executed.

laf commented:
This good for a merge now @adaniels21487?

@librenms/reviewers ?

adaniels21487 commented:
Thanks @laf I will log a new issue for the PHP on update..

On this one, I have added most of @paulgear suggestions:

Fixed:
- allow editing & deleting services from the master list of services
- when you click the name of the device in the service list, it should take you to the service screen for that device, not the overview screen
- The renames of discover_service() to service_discover() (and the other requivalents) are counter-intuitive; function names should be or begin with verbs.
- We should fix "TODO: Use proper Nagios service status. 0=Ok,1=Warning,2=Critical,Else=Unknown" before we release this (this was one of my pet peeves with the current services code). The values in use by the state management stuff that @Rosiak has been working on are modelled directly on Nagios, and we should integrate with that. Not many people are doing service polling now, and we should make the switch sooner rather than later, because this PR will cause a lot more people to be interested in service alerting. A 1-2 week lead time should be enough for people to get their alert rules fixed.

Future Work:
- would be nice if the service list were sorted; I had a branch ready to go with this but decided to wait until this PR is merged before submitting
- Eliminate the extra files in includes/services/check_*.inc.php and just use a config array (or better still, database entry). I think we should make this as configurable as possible by the end user. Ideally, when we add a service check, we should default to using the parameters which were used when the last check of the same type was added, so that people can just use any Nagios check they want without us having to explicitly support it. (Maybe we could allow a custom path for the check rather than requiring them all to be in one place?)

I can't find the display bug anywhere. can someone who is seeing it send me their RRD command?
Otherwise, I think its ready for a merge, or if you prefer I can squash to a single commit for neatness?

adaniels21487 commented:
bump :)

laf commented:
If you can rebase and take a look at https://scrutinizer-ci.com/g/librenms/librenms/inspections/454d5e67-fc81-4c08-820b-a0a18c86f5f8/issues/files/includes/services.inc.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

Will merge after that.

adaniels21487 commented:
Hi @laf 
All fixed.

f0o commented:
Looks good to me

jcdouglas commented:
Not sure if this is the appropriate venue or if I should create a new issue apologies, as I did do a search, but I ran into an issue with check-services and "check_dhcp":

`RRD[update /opt/librenms/rrd/my.server.lan/services-8.rrd N:-0.002345]
Nagios Service - 9
Request:  /usr/lib/nagios/plugins/check_dhcp -H ns.section31.lan -r
/usr/lib/nagios/plugins/check_dhcp: invalid option -- 'H'
Perf Data - None.
Response: Usage:
 check_dhcp [-v] [-u] [-s serverip] [-r requestedip] [-t timeout]
                  [-i interface] [-m mac]
OK u:0.00 s:0.03 r:0.26
`
Not sure if this is with check-services.php as the other plugins I've added so far use the '-H' option and appears to work great (thanks developers), or if it's a bug with the nagios plugin 'check_dhcp'.

adaniels21487 commented:
Hi @jcdouglas
The default command inserts a -H. I didnt want to pull this out as it would break things.
Try this

Create a new file:

    /opt/librenms/includes/services/check_dhcp.inc.php

With this content:

```php
<?php
$check_cmd = $config['nagios_plugins'] . "/check_dhcp ".$service['service_param'];
```

Create or edit your dhcp check so that you put the check arguments in the parameters field, The IP address will not be used.

Let me know if this works for you and I'll create a new PR.

Aaron

jcdouglas commented:
Yes, this worked.  Thank you so much.

On Thu, Apr 14, 2016 at 4:56 PM, Aaron Daniels <notifications@github.com>
wrote:

> Hi @jcdouglas <https://github.com/jcdouglas>
> The default command inserts a -H. I didnt want to pull this out as it
> would break things.
> Try this
>
> Create a new file:
>
> /opt/librenms/includes/services/check_dhcp.inc.php
>
> With this content:
>
> <?php$check_cmd = $config['nagios_plugins'] . "/check_dhcp ".$service['service_param'];
>
> Create or edit your dhcp check so that you put the check arguments in the
> parameters field, The IP address will not be used.
>
> Let me know if this works for you and I'll create a new PR.
>
> Aaron
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/librenms/librenms/pull/3244#issuecomment-210145913>
>
